docs/source/user_guide.md

```diff
@@ -153,8 +165,190 @@ name: new-project
+I'll add a variable `FOO` with the default value `has-default-value` and the variable `BAR` with
+no default value.
+
+```yaml
+name: new-project
+environments:
+  default:
+    - environment.yml
+
+variables:
+  FOO: has-default-value
+  BAR:
+
+commands: {}
+```
+
+The values of the variables can be set or overridden by the use of a `.env` file (see the
+[python-dotenv documentation](https://saurabh-kumar.com/python-dotenv/#file-format) for more details) or
+by setting the variable in your shell.
+
+In this example the `BAR` environment variable is unset so `conda project activate` will not start
+unless a value is provided by the shell or in a `.env` file.
+
+```
+❯ conda project activate
+CondaProjectError: The following variables do not have a default value and values
+were not provided in the .env file or set on the command line when executing 'conda project run':
+BAR
+```
+
+On Unix you can do the following
+
+```
+❯ BAR=set-on-cli conda project activate
+## Project environment default activated in a new shell.
+## Exit this shell to de-activate.
+❯ conda activate /Users/adefusco/Development/conda-incubator/conda-project/examples/p/envs/default
+```
+
+On Windows in either `cmd.exe` you can use `set variable=value`
+
+```
+> set BAR=set-in-shell
+> conda project activate
+```
+
+and finally, in Powershell you would use `$env`
+
+```
+> $env:BAR = 'set-in-shell'
+> conda project activate
+```
+
+## Defining and running commands
+
+Conda Project supports running commands as if the desired environment were activated, similar
+to [conda run](https://docs.conda.io/projects/conda/en/latest/commands/run.html).
+
+In Conda Project the `run` command can be used to execute ad-hoc commands or those defined
+in the `conda-project.yml` file. Note that `conda project run` should not be utilized after
+`conda project activate`. The `run` command will not work from within an activated environment.
+
+Let's start with a simple python script called `vars.py`, that prints environment variables
+and optionally takes an argument `--version` to also print the Python version.
+
+```python
+import os
+import sys
+
+if len(sys.argv) > 1 and sys.argv[1] == '--version':
+    print("0.0.1")
+    sys.exit(1)
+else:
+    print(f"The value of FOO is {os.environ.get('FOO')}")
+    print(f"The value of BAR is {os.environ.get('BAR')}")
+    print(f"The value of BAZ is {os.environ.get('BAZ')}")
 ```
 
+Note that when the `--version` flag is provided the return code for this script is 1, meaning
+a failure.
+
+We can execute this script with the `default` environment as an ad-hoc command using the Python
+interpreter provided by the default environment. Here a `.env` file is utilized to provide the
+value of the BAR variable.
+
+```
+> BAR=set-on-cli conda project run python vars.py
+The value of FOO is has-default-value
+The value of BAR is set-on-cli
+The value of BAZ is None
+```
+
+On Linux, Mac, and Windows the return code of the `conda project run` is set to the
```

Comment:
Nice!

Never tried this with anaconda project but this sounds like a (useful!) new feature to me!

Reply:
This is how anaconda-project run works, too 😀